torch/distributed/fsdp/flat_param.py

```diff
@@ -575,6 +585,415 @@ def shard_metadata(
             self.flat_param._shard_param_offsets[:],  # type: ignore[attr-defined]
         )
 
+    ###################
+    # UNSHARD/RESHARD #
+    ###################
+    def pre_unshard(self) -> bool:
```

Comment:
To accommodate the pre-unshard stream, `pre_unshard()` returns whether or not it was a no-op.